File: Tutorials/Codes/System_Dynamics_Course/Chapter14/Lesson5/Chapter14_Lesson5.cpp
// Chapter14_Lesson5.cpp
// Limit Cycles, Multiple Equilibria, and Basic Bifurcation Notions
// C++17: RK4 simulation of Van der Pol oscillator and equilibrium classification
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <fstream>
#include <string>

struct State {
    double x;
    double y;
};

State vdp_rhs(const State& s, double mu) {
    State ds;
    ds.x = s.y;
    ds.y = mu * (1.0 - s.x * s.x) * s.y - s.x;
    return ds;
}

State add(const State& a, const State& b, double scale=1.0) {
    return {a.x + scale * b.x, a.y + scale * b.y};
}

State rk4_step(const State& s, double h, double mu) {
    State k1 = vdp_rhs(s, mu);
    State k2 = vdp_rhs(add(s, k1, 0.5*h), mu);
    State k3 = vdp_rhs(add(s, k2, 0.5*h), mu);
    State k4 = vdp_rhs(add(s, k3, h), mu);
    State out;
    out.x = s.x + (h/6.0) * (k1.x + 2.0*k2.x + 2.0*k3.x + k4.x);
    out.y = s.y + (h/6.0) * (k1.y + 2.0*k2.y + 2.0*k3.y + k4.y);
    return out;
}

double estimate_period(const std::vector<double>& t, const std::vector<State>& z) {
    std::vector<double> crossings;
    for (size_t k = 0; k + 1 < z.size(); ++k) {
        if (z[k].y < 0.0 && z[k+1].y >= 0.0) {
            double denom = (z[k+1].y - z[k].y);
            if (std::abs(denom) < 1e-14) continue;
            double alpha = -z[k].y / denom;
            double tc = t[k] + alpha * (t[k+1] - t[k]);
            double xc = z[k].x + alpha * (z[k+1].x - z[k].x);
            if (xc > 0.0) crossings.push_back(tc);
        }
    }
    if (crossings.size() < 3) return std::nan("");
    size_t start = (crossings.size() > 5) ? crossings.size() - 5 : 1;
    double sum = 0.0;
    int cnt = 0;
    for (size_t i = start; i < crossings.size(); ++i) {
        sum += (crossings[i] - crossings[i-1]);
        cnt++;
    }
    return (cnt > 0) ? sum / cnt : std::nan("");
}

std::string stability_pitchfork(double r, double xeq) {
    double lambda = r - 3.0 * xeq * xeq;
    if (lambda < 0.0) return "stable";
    if (lambda > 0.0) return "unstable";
    return "nonhyperbolic";
}

std::string stability_saddlenode(double xeq) {
    double lambda = -2.0 * xeq;
    if (lambda < 0.0) return "stable";
    if (lambda > 0.0) return "unstable";
    return "nonhyperbolic";
}

int main() {
    const double mu = 1.0;
    const double h = 0.01;
    const double T = 80.0;
    const int N = static_cast<int>(T / h);

    std::vector<double> t(N + 1);
    std::vector<State> z(N + 1);
    z[0] = {2.0, 0.1};
    t[0] = 0.0;

    for (int k = 0; k < N; ++k) {
        t[k+1] = t[k] + h;
        z[k+1] = rk4_step(z[k], h, mu);
    }

    // Write phase-portrait data to CSV for plotting externally
    std::ofstream csv("Chapter14_Lesson5_vdp_phase.csv");
    csv << "t,x,y\n";
    for (int k = 0; k <= N; ++k) {
        csv << t[k] << "," << z[k].x << "," << z[k].y << "\n";
    }
    csv.close();

    // Estimate asymptotic period from latter part
    std::vector<double> t_tail(t.begin() + N/2, t.end());
    std::vector<State> z_tail(z.begin() + N/2, z.end());
    double period = estimate_period(t_tail, z_tail);

    double amp = 0.0;
    for (size_t k = N/2; k < z.size(); ++k) amp = std::max(amp, std::abs(z[k].x));

    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Van der Pol (mu=" << mu << ") limit-cycle estimate\n";
    std::cout << "Amplitude |x|_max ~= " << amp << "\n";
    std::cout << "Period ~= " << period << "\n\n";

    // Print equilibrium branches for two normal forms (sampled values)
    std::cout << "Pitchfork normal form: xdot = r*x - x^3\n";
    for (double r : {-1.0, 0.0, 1.0}) {
        std::cout << "r = " << r << ": ";
        std::cout << "x*=0 (" << stability_pitchfork(r, 0.0) << ")";
        if (r >= 0.0) {
            double x1 = std::sqrt(r), x2 = -std::sqrt(r);
            std::cout << ", x*=+" << x1 << " (" << stability_pitchfork(r, x1) << ")";
            std::cout << ", x*=" << x2 << " (" << stability_pitchfork(r, x2) << ")";
        }
        std::cout << "\n";
    }

    std::cout << "\nSaddle-node normal form: xdot = r - x^2\n";
    for (double r : {-1.0, 0.0, 1.0}) {
        std::cout << "r = " << r << ": ";
        if (r < 0.0) {
            std::cout << "no real equilibria\n";
        } else if (r == 0.0) {
            std::cout << "x*=0 (nonhyperbolic)\n";
        } else {
            double x1 = std::sqrt(r), x2 = -std::sqrt(r);
            std::cout << "x*=+" << x1 << " (" << stability_saddlenode(x1) << "), ";
            std::cout << "x*=" << x2 << " (" << stability_saddlenode(x2) << ")\n";
        }
    }

    std::cout << "\nCSV written: Chapter14_Lesson5_vdp_phase.csv\n";
    return 0;
}

File: Tutorials/Codes/System_Dynamics_Course/Chapter14/Lesson5/Chapter14_Lesson5.java
// Chapter14_Lesson5.java
// Limit Cycles, Multiple Equilibria, and Basic Bifurcation Notions
// Java implementation: RK4 simulation of Van der Pol + basic bifurcation classifications
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Chapter14_Lesson5 {
    static class State {
        double x, y;
        State(double x, double y) { this.x = x; this.y = y; }
    }

    static State rhsVanDerPol(State s, double mu) {
        double dx = s.y;
        double dy = mu * (1.0 - s.x * s.x) * s.y - s.x;
        return new State(dx, dy);
    }

    static State add(State a, State b, double scale) {
        return new State(a.x + scale * b.x, a.y + scale * b.y);
    }

    static State rk4Step(State s, double h, double mu) {
        State k1 = rhsVanDerPol(s, mu);
        State k2 = rhsVanDerPol(add(s, k1, 0.5 * h), mu);
        State k3 = rhsVanDerPol(add(s, k2, 0.5 * h), mu);
        State k4 = rhsVanDerPol(add(s, k3, h), mu);
        return new State(
            s.x + (h / 6.0) * (k1.x + 2*k2.x + 2*k3.x + k4.x),
            s.y + (h / 6.0) * (k1.y + 2*k2.y + 2*k3.y + k4.y)
        );
    }

    static String pitchforkStability(double r, double xeq) {
        double lambda = r - 3.0 * xeq * xeq;
        if (lambda < 0) return "stable";
        if (lambda > 0) return "unstable";
        return "nonhyperbolic";
    }

    static String saddleNodeStability(double xeq) {
        double lambda = -2.0 * xeq;
        if (lambda < 0) return "stable";
        if (lambda > 0) return "unstable";
        return "nonhyperbolic";
    }

    static double estimatePeriod(List<Double> t, List<State> z) {
        List<Double> crossings = new ArrayList<>();
        for (int k = 0; k < z.size() - 1; k++) {
            if (z.get(k).y < 0.0 && z.get(k+1).y >= 0.0) {
                double denom = z.get(k+1).y - z.get(k).y;
                if (Math.abs(denom) < 1e-14) continue;
                double alpha = -z.get(k).y / denom;
                double tc = t.get(k) + alpha * (t.get(k+1) - t.get(k));
                double xc = z.get(k).x + alpha * (z.get(k+1).x - z.get(k).x);
                if (xc > 0) crossings.add(tc);
            }
        }
        if (crossings.size() < 3) return Double.NaN;
        int start = Math.max(1, crossings.size() - 5);
        double sum = 0.0;
        int count = 0;
        for (int i = start; i < crossings.size(); i++) {
            sum += crossings.get(i) - crossings.get(i - 1);
            count++;
        }
        return (count > 0) ? sum / count : Double.NaN;
    }

    public static void main(String[] args) throws IOException {
        double mu = 1.0;
        double h = 0.01;
        double T = 80.0;
        int N = (int)(T / h);

        List<Double> t = new ArrayList<>(N + 1);
        List<State> z = new ArrayList<>(N + 1);
        t.add(0.0);
        z.add(new State(2.0, 0.1));

        for (int k = 0; k < N; k++) {
            t.add(t.get(k) + h);
            z.add(rk4Step(z.get(k), h, mu));
        }

        double amp = 0.0;
        for (int k = N/2; k < z.size(); k++) {
            amp = Math.max(amp, Math.abs(z.get(k).x));
        }

        List<Double> tTail = t.subList(N/2, t.size());
        List<State> zTail = z.subList(N/2, z.size());
        double period = estimatePeriod(tTail, zTail);

        try (FileWriter fw = new FileWriter("Chapter14_Lesson5_vdp_phase.csv")) {
            fw.write("t,x,y\n");
            for (int k = 0; k < z.size(); k++) {
                fw.write(t.get(k) + "," + z.get(k).x + "," + z.get(k).y + "\n");
            }
        }

        System.out.println("Van der Pol limit-cycle summary");
        System.out.printf("mu = %.3f%n", mu);
        System.out.printf("Amplitude |x|_max ~= %.6f%n", amp);
        System.out.printf("Period ~= %.6f%n%n", period);

        double[] rVals = {-1.0, 0.0, 1.0};
        System.out.println("Pitchfork normal form: xdot = r*x - x^3");
        for (double r : rVals) {
            System.out.print("r = " + r + ": x*=0 (" + pitchforkStability(r, 0.0) + ")");
            if (r >= 0.0) {
                double x1 = Math.sqrt(r), x2 = -Math.sqrt(r);
                System.out.print(", x*=+" + x1 + " (" + pitchforkStability(r, x1) + ")");
                System.out.print(", x*=" + x2 + " (" + pitchforkStability(r, x2) + ")");
            }
            System.out.println();
        }

        System.out.println("\nSaddle-node normal form: xdot = r - x^2");
        for (double r : rVals) {
            System.out.print("r = " + r + ": ");
            if (r < 0.0) {
                System.out.println("no real equilibria");
            } else if (r == 0.0) {
                System.out.println("x*=0 (nonhyperbolic)");
            } else {
                double x1 = Math.sqrt(r), x2 = -Math.sqrt(r);
                System.out.println("x*=+" + x1 + " (" + saddleNodeStability(x1) + "), x*=" + x2 + " (" + saddleNodeStability(x2) + ")");
            }
        }

        System.out.println("\nCSV written: Chapter14_Lesson5_vdp_phase.csv");
    }
}
